Update getLastValue docs (#238)
openDAQ/openDAQ: docs/Antora/modules/howto_guides/pages/howto_measure_single_value.adoc

= Measure a single value

openDAQ offers a variety of solutions for reading Signal values, including Readers. 

However, for developers requiring access to only the most recent value, the Signal interface provides a method for doing just that.

[#last_value_signal]
== Getting the last value of a Signal

Reading the last value of a Signal yields an openDAQ Base object-type value. Said value contains the last sample of the most recent Packet sent through the Signal.

[tabs]
====
Cpp::
+
[source,cpp]
----
// Get last value of a Signal
auto myLastValue = mySignal.getLastValue();
----
Python::
+
[source,python]
----
# Get last value of a Signal
my_last_value = my_signal.last_value
----
====

[#last_value_signal_types]
== Last value and Signal types

> The method for getting the last value yields an openDAQ Base object-type value.

Reviewer: A 1-line code example of how to retrieve the value type would be great here:

```
SampleType type = signal.getDescriptor().getSampleType();
```

The underlying type of the `IBaseObject` we get from our `getLastValue` call can be cast into either:

* `IFloat` if the type is `Float32` or `Float64`
* `IInteger` if the type is `Int8` through `Int64` or `UInt8` through `UInt64`
* `IComplexNumber` if the type is `ComplexFloat32` or `ComplexFloat64`
* `IRange` if the type is `RangeInt64`
* `IStruct` if the type is `Struct`
* `IList` of the aforementioned types if there is exactly one dimension in the single value

An example of casting the `lastValue` object into an `IFloat` object:

[tabs]
====
Cpp::
+
[source,cpp]
----
// Cast to IFloat
auto number = lastValue.asPtr<IFloat>();
----
Python::
+
[source,python]
----

----
====

In the above example, the Data Descriptor of the Data Packet or its Signal on which `getLastValue` (`last_value` in Python) is called might be:

[tabs]
====
Cpp::
+
[source,cpp]
----
// The Data Descriptor for Float64
auto descriptor = DataDescriptorBuilder().setSampleType(SampleType::Float64).build();
----
Python::
+
[source,python]
----

----
====

[#last_value_range]

> == Getting the last value with a Range

Reviewer: I'd omit this section. We just need a list of type conversions at the top for users to refer to. Beyond that they should understand how to use certain openDAQ types - this should be a separate guide if we need it.

The following example demonstrates casting the `lastValue` object into an `IRange`. Something usefeul we can then do is call the methods of the underlying type and access its members, such as the `getLowValue` (`low_value` in Python) and the `getHighValue` (`high_value` in Python), in this case provided by the `IRange` interface.

[tabs]
====
Cpp::
+
[source,cpp]
----
// Cast to IRange
auto range = lastValue.asPtr<IRange>();
// Extract values
auto low = range.getLowValue();
auto high = range.getHighValue();
----
Python::
+
[source,python]
----

----
====

In the above example, the Data Descriptor of the Data Packet or its Signal on which `getLastValue` (`last_value` in Python) is called is similar to the the one from the previous example (the Data Descriptor with Sample Type `Float64`), except the Sample Type is `RangeInt64`.

[#last_value_complex_number]

> == Getting the last value with a Complex Number

Reviewer: As above, I don't think an explanation of how complex numbers work is needed in this how-to

In the following example:

* Last value is extracted via a `getLastValue` (`last_value` in Python) call
* `lastValue` object is cast into a `IComplexNumber` type (not needed in Python)
* Methods of the underlying type such as the `getReal` (`real` in Python) and the `getImaginary` (`imaginary` in Python) are called, in this case provided by the `IComplexNumber` interface

[tabs]
====
Cpp::
+
[source,cpp]
----
// Get last value of a Signal
auto lastValue = signal.getLastValue();
// Cast to IComplexNumber
auto complex = lastValue.asPtr<IComplexNumber>();
// Extract values
auto real = complex.getReal();
auto imaginary = complex.getImaginary();
----
Python::
+
[source,python]
----

====

A Data Descriptor for this Complex Number example is again very similar to the ones from previous examples, except the Sample Type might be `ComplexFloat32` or `ComplexFloat64` or  instead of `Float64` or `RangeInt64`.

[#last_value_struct]
== Getting the last value with a Struct

> A Data Descriptor for a Struct might be:

Reviewer: We should simply describe the descriptor and how the signal in question is structured. No need to provide the builder code sample.

----
Python::
+
[source,python]
----

----
====

In the above example, a Data Descriptor of a custom Struct that has two fields is first built. The first field has Sample Type `Int32` and the second one has Sample Type `Float64`.

[CAUTION]
====
`opendaq.SampleType.Struct` is not yet supported in Python.
====

[NOTE]
====
Because our Struct represents a custom type, we must call `setDescriptor` (`descriptor` in Python) on our Signal in order to add `MyStruct` to the Type Manager.
====

[CAUTION]
====
All custom Structs and their respective fields must have their names set, which is a deviation from the previous examples. In the above code this is done by the calls to `setName` (`name` in Python) method of the Data Descriptor Builder.
====

After some data has been sent to our Signal, we can capture a single value similar to the previous examples:

[tabs]
====
Cpp::
+
[source,cpp]
----
// Get last value of a Signal
auto lastValue = signal.getLastValue();
// Cast to IStruct
auto myStruct = lastValue.asPtr<IStruct>();
// Extract values
auto myInt = myStruct.get("MyInt32");
auto myFloat = myStruct.get("MyFloat64");
----
Python::
+
[source,python]
----

----
====

[CAUTION]
====

> If we want to call `getLastValue` (`last_value` in Python) on a Data Packet (as opposed to a Signal) and the underlying data type is a Struct, we must provide the optional parameter, Type Manager, which has had the custom Struct added to it.

Reviewer: Honestly, I would not even mention that we can read the last value on a packet. The guide aims to describe how to read the last value of a signal, the packet last value is mostly for internal usage and will confuse readers.

= Measure a single value

OpenDAQ offers a variety of solutions for reading Signal values, including Readers. 

For developers requiring access to only the most recent value, the Signal and Data Packet interfaces provide a method named `getLastValue` (`last_value` in Python).

[#last_value_signal]
== Getting the last value of a Signal

When the `getLastValue` method (`last_value` in Python)  is called on a Signal, it yields an `IBaseObject` containing the last value.

[tabs]
====
Cpp::
+
[source,cpp]
----
// Get last value of a Signal
auto lastValue = signal.getLastValue();
----
Python::
+
[source,python]
----
# Get last value of a Signal
last_value = signal.last_value
----
====

[#last_value_data_packet]
== Getting last value of a Data Packet

Similar is true when calling the `getLastValue` method (`last_value` in Python)  on a Data Packet.

[tabs]
====
Cpp::
+
[source,cpp]
----
// Get last value of a Data Packet
auto lastValue = packet.getLastValue();
----
Python::
+
[source,python]
----
# Get last value of a Data Packet

> packet.last_value()  # TODO ()???

Author: Not sure if brackets are appropriate here... It's suppored to take one argument for Struct and the interpreter suggested it. Couldn't test Python code...

> packet.last_value()  # TODO ()???

Author: Not sure if brackets are appropriate here... It's supported to take one argument for Struct and the interpreter suggested it. Couldn't test Python code (however I have some snippets in a separate branch)...

----
====

[#printing_the_last_value]
== Printing the last value

No matter the source (either the Signal or the Data Packet), the resulting variable can be, for example, printed into the console.

[tabs]
====
Cpp::
+
[source,cpp]
----
// Print last value
std::cout << "lastValue: " << lastValue << std::endl;
----
Python::
+
[source,python]
----
# Print last value
print("last value: ", last_value)
----
====

[#casting_the_last_value]
== Casting the last value and using the cast variable

> In C++, if we would like to do something useful with the `IBaseObject`, such as access its underlying methods, we need to cast it into the correct type first. This is not necessary in Python, however.

Reviewer 2: ```suggestion
In C++, the function `getLastValue` returns an object of type IBaseObject. To manipulate it, developers must explicitly convert it to the appropriate type. However, in Python, this conversion process is handled automatically.
```

The underlying type of the `IBaseObject` we get from our `getLastValue` call can be cast into either `IFloat` if the type is `Float32` or `Float64`, to `IInteger` if the type is `Int8` through `Int64` or `UInt8` through `UInt64`, to `IComplexNumber` if type is `ComplexFloat32` or `ComplexFloat64`, to `IRange` if the type is `RangeInt64`, to `IStruct` if the type is `Struct`, and to `IList` of the aforementioned types if there is exactly one dimension in the single value.

> In this example, we cast `lastValue` object into an `IFloat` like so:

Reviewer 2: ```suggestion
example of casting the `lastValue` object into an `IFloat` object:
```

[tabs]
====
Cpp::
+
[source,cpp]
----
// Cast to IFloat
auto number = lastValue.asPtr<IFloat>();
----
====

> In the above example, the Data Descriptor of the Data Packet or its Signal on which we call `getLastValue` (`last_value` in Python) is:

Reviewer 2: remove `we`

[tabs]
====
Cpp::
+
[source,cpp]
----
// The Data Descriptor for Float64
auto descriptor = DataDescriptorBuilder().setSampleType(SampleType::Float64).build();
----
Python::
+
[source,python]
----
# The Data Descriptor for Float64
builder = opendaq.DataDescriptorBuilder()
builder.sample_type = opendaq.SampleType.Float64
descriptor = builder.build()
----
====

[#last_value_range]
== Getting the last value with a Range

> In the following example, we cast the `lastValue` object into an `IRange` like so:

Reviewer 2: ```suggestion
The following example demonstrate casting the `lastValue` object into an `IRange`
```

[tabs]
====
Cpp::
+
[source,cpp]
----
// Cast to IRange
auto range = lastValue.asPtr<IRange>();
----
====

An example of something we could now do is call the methods of the underlying type and access its members such as the `getLowValue` (`low_value` in Python) and the `getHighValue` (`high_value` in Python), in this case provided by the `IRange` interface.

> [tabs]

Reviewer 2: i would squash this example with previous one

====
Cpp::
+
[source,cpp]
----
// Extract values
auto low = range.getLowValue();
auto high = range.getHighValue();
----
Python::
+
[source,python]
----
# Extract values
low = last_value.low_value
high = last_value.high_value
----
====

> In the above example, the Data Descriptor of the Data Packet or its Signal on which we call `getLastValue` (`last_value` in Python) is:

Reviewer 2: remove 'we'

> [tabs]

Reviewer 2: i would remove this example and add comment that we are assuming that type of data descriptor is int64

====
Cpp::
+
[source,cpp]
----
// The Data Descriptor for RangeInt64
auto descriptor = DataDescriptorBuilder().setSampleType(SampleType::RangeInt64).build();
----
Python::
+
[source,python]
----
# The Data Descriptor for RangeInt64
builder = opendaq.DataDescriptorBuilder()
builder.sample_type = opendaq.SampleType.RangeInt64
descriptor = builder.build()
----
====

[#last_value_complex_number]
== Getting the last value with a Complex Number

> In the following example, we first extract the last value via `getLastValue` (`last_value` in Python) call, then we cast the `lastValue` object into a `IComplexNumber` type (not needed in Python). Lastly, call the methods of the underlying type such as the `getReal` (`real` in Python) and the `getImaginary` (`imaginary` in Python), in this case provided by the `IComplexNumber` interface.

Reviewer 2: - remove `we`
- you can make a list of steps (up ur choice )

[tabs]
====
Cpp::
+
[source,cpp]
----
// Get last value of a Signal
auto lastValue = signal.getLastValue();
// Cast to IComplexNumber
auto complex = lastValue.asPtr<IComplexNumber>();
// Extract values
auto real = complex.getReal();
auto imaginary = complex.getImaginary();
----
Python::
+
[source,python]
----
# Get last value of a Signal
last_value = signal.last_value
# Extract values
real = last_value.real
imaginary = last_value.imaginary
====

> A Data Descriptor for a Complex Number might be:

Reviewer 2: just add assumption in prev exmaple. and remove this example

----
====

[#last_value_struct]
== Getting the last value with a Struct

A Data Descriptor for a Struct might be:

[tabs]
====
Cpp::
+
[source,cpp]
----
// Create a Data Descriptor
auto descriptor = DataDescriptorBuilder()
					  .setName("MyStruct")
					  .setSampleType(SampleType::Struct)
					  .setStructFields(List<DataDescriptorPtr>(
						  DataDescriptorBuilder()
							  .setName("MyInt32")
							  .setSampleType(SampleType::Int32)
							  .build(),
						  DataDescriptorBuilder()
							  .setName("MyFloat64")
							  .setSampleType(SampleType::Float64)
							  .build()))
					  .build();
// Set the Data Descriptor, thereby adding MyStruct to the Type Manager
signal.setDescriptor(descriptor);
----
Python::
+
[source,python]
----
# Create a Data Descriptor
builder_int = opendaq.DataDescriptorBuilder()
builder_int.name = "MyInt32"
builder_int.sample_type = opendaq.SampleType.Int32
descriptor_int = builder_int.build()

builder_float = opendaq.DataDescriptorBuilder()
builder_float.name = "MyFloat64"
builder_float.sample_type = opendaq.SampleType.Float64
descriptor_float = builder_float.build()

builder = opendaq.DataDescriptorBuilder()
builder.name = "MyStruct"

> builder.sample_type = opendaq.SampleType.Struct  # TODO not implemented

Author: SampleType.Struct is not supported in Python yet.

Reviewer 2: add note in doc

struct_fields = opendaq.List()
struct_fields.append(descriptor_int)
struct_fields.append(descriptor_float)
builder.struct_fields = struct_fields
descriptor = builder.build()

# Set the Data Descriptor, thereby adding MyStruct to the Type Manager

> signal.descriptor = descriptor  # TODO setter not defined

Author: This setter doesn't seem to be defined yet.

Reviewer 2: setter in ISignalConfig

----
====

> In the above example, we first build a Data Descriptor of a custom Struct that has two fields. The first field has Sample Type `Int32` and the second one has Sample Type `Float64`.

Reviewer 2: remove 'we'
